llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4106,15 +4133,11 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
     if (Rem->isZero()) {
       // Accept MaxFixedVF if we do not have a tail.
       LLVM_DEBUG(dbgs() << "LV: No tail will remain for any chosen VF.\n");
+      disableTailFolding();
```

Comment:
This is needed, because we need to first need to compute the max VFs, assuming tail folding, right?

Could we end up in a scenario, where we MaxSafeNumberOfElements is 3, we max scalable VF of 4 is picked and then tail-folding is disabled here and no EVL will be used, vectorizing incorrectly with VF 4?

Reply:
1. Right.
2. Yes, looks so. Need to do extra processing here

Reply:
1. Added the extra check above, when setting MaxPowerOf2RuntimeVF
2. I'm afraid there might be some side effects, if we'll keep tail-folding mode ON, while it is OFF. Better explicitly set it to OFF, I think

Reply:
> 2. I'm afraid there might be some side effects, if we'll keep tail-folding mode ON, while it is OFF. Better explicitly set it to OFF, I think

Agreed, I was suggesting if it is possible to keep `setTailFoldingStyles` at its original place and then try to maximize MaxScalableVF below where we already deal with the EVL case ( `if (getTailFoldingStyle() == TailFoldingStyle::DataWithEVL) {` below)

Reply:
Maybe it can be done right now, but the following patch still requires moving it up (for-non-power-2 support)

Reply:
If possible to keep it in the original place without much extra work that would be preferable, moving when actually needed?